More details on metadata adversaries (#25)
phnx-im/docs: src/threat_model/qualitative/metadata.md

# Metadata

Protection of metadata is a key goal of the Phoenix homeserver protocol. Metadata typically includes information like the sender and recipient of a message, the time it was sent, and the size of the message, and the group conversation in which it was sent. Metadata aggregation can reveal patterns of communication and can be used to infer social relationships. In particular, correlation can be used to infer social graphs.

The metadata threat model considers two types of adversaries: snapshot adversaries and active observer adversaries. This section describes the two adversaries and the information they have access to.

## Snapshot adversarires

> Snapshot adversaries have access to snapshots of the server's persisted state. They can see the metadata of all messages that have been sent and received up to the point of the snapshot and that haven't been deleted from the long-term storage. An example of a snapshot adversary is a government that has seized the server's hard drives, or a provider that is forced to comply with a government request to hand over data.

Reviewer: ```suggestion
Snapshot adversaries have access to snapshots of the server's persisted state. They can see the metadata of all messages sent and received up to the point of the snapshot that haven't been deleted from the long-term storage. An example of a snapshot adversary is law enforcement that has seized the server's hard drives or a provider forced to comply with a legal request to hand over data.
```

> The main protection against the snapshot adversary is encryption at rest for most relevant metadata. In this instance, encryption at rest means that clients hold the key material and supply it to the server for individual queries. The server can decrypt the data and process the query before re-encrypting the data and deleting the key from memory.

Reviewer: ```suggestion
The main protection against the snapshot adversary is encryption at rest for most relevant metadata. In this instance, encryption at rest means that clients hold the key material and supply it to the server for individual queries. The server can decrypt the data and process the query before re-encrypting the data and deleting the key from memory.
The underlying assumption is that the server infrastructure cannot record the keys without substantial technical changes to the infrastructure. Such technical changes are out-of-scope for this kind of adversary.
```

As a consequence, snapshots become largely useless to the adversary since the senders of messages, the group in which messages were sent, etc. are encrypted. The encryption keys are either only held by clients, or are only held temporarily by the server in volatile memory and are therefore not part of the snapshot.

### Data available to snapshot adversaries

Despite the use of encryption at rest, the snapshot adversary can still see the following data:

- For each group on the DS

>   - The rough size of the group

Reviewer: We should mention that this depends mostly (only?) on the padding size

>   - The timestamp of the last time the group was active (e.g. due to a key update)

Reviewer: This is not systematic and could be coarse, might be worth mentioning that

> - Reserved group IDs

Reviewer: This needs context: Group IDs are random identifiers that are not correlated with anything

- AS key material
  - Credentials of all clients (includes usernames)
  - AS credentials and key material
  - OPAQUE key material
  - Privacy Pass key material

> - The number of messages in the connection queues of all clients (but not the contents of the messages)

Reviewer: Worth noting that clients are not correlated with users

Author: Clients are still correlated with users. However, there are also the aliases and those are not correlated.

- All user pseudonyms and their associated client pseudonyms
- All currently available KeyPackages of all clients
- The number of messages in all client queues and the (padded) length of each message (but not the contents)
- QS key material
  - QS signing keys

>   - QS HPKE encryption keys

Reviewer: In general, this list needs more context regarding correlation (or the lack thereof). We shouldn't assume readers understand that e.g. a client is not tied to an AS user on the server.

## Active observer adversaries

Active observer adversaries have full access to the server including its volatile memory in addition to the snapshots.

The main protection against the active observer is that users interact with the services using per-group pseudonyms instead of their real identities. This means that the adversary cannot link any individual action of a user with the user's identity as seen by other users. The adversary can still see the metadata of messages, but cannot link them to the real identities of the users.

### Data available to active observer adversaries

As per definition, active observer adversaries have access to all data on the server at the time of their activity.

> 

Reviewer: We kind of have the same definition twice. It's also not clear if the active observer can manipulate data. On one hand, this is implied but the word 'active', on the other hand it's excluded by the word 'observer'. We should probably discuss this again.

Reviewer 2: As an inspiration: In other areas, "honest", "curious", and "malicious" are terms to identify fully passive, observant, or actively manipulating threat actors.

Reviewer: I guess "curious" would be the one here.

Author: I looked around a bit an settled on honest-but-curious, which I have seen used before and which seems to fit quite well for our adversarial model.

## Traffic analysis

> Traffic analysis, i.e. the analysis of data traffic patterns, is not considered part of the threat model. While the use of network metadata can allow an active observer (be it on the server or just the network) to infer social graphs and other sensitive information, the Phoenix homeserver protocol does not aim to protect against such attacks. Instead, common countermeasures such as onion routing or the use of mixnets can be used in conjunction with the Phoenix homeserver protocol to mitigate such attacks. However, to the extent possible, the protocol avoids behaviours obvious traffic patterns such as batched queries, messages at regular intervals, etc.

Reviewer: "behaviours obvious" <- get rid of one of these words

#### Users, clients and aliases (AS)

With a snapshot, the SA obtains a list of every user registered with the AS, as well as of their clients, including each client's public authentication key material (i.e. client credentials). The SA can also view a list of all aliases registered with the AS. Note that there is no link between aliases and user identities.

For both clients and aliases, the SA can view the connection queues, as well as the number of currently enqueued messages within them (but not the content, as connection requests are encrypted).

The SA also has access to the key material pre-published for connection establishment (connection packages) of all clients.

### Data not available to snapshot adversaries

> Generally, the SA only has access to the data listed above. In particular, the SA does not have access to

Reviewer: Just a nit, but the word "generally" introduces some fuzziness we don't want.

- push tokens

> - group memberhsip

Reviewer: ```suggestion
- group membership
```

- links between identities and pseudonyms
- links between identities and aliases

#### Client pseudonyms (QS)

On the QS, the SA can observe all user pseudonyms and their associated client pseudonyms. Note that these pseudonyms are not linked to any aliases, client- or user identities.

> For each client pseudonym the SA can see pre-published key material (add packages) and queues. The SA can thus also see the number of messages in each queue, but not their content due to at-rest encryption.

Reviewer: ```suggestion
The SA can see pre-published key material (add packages) and queues for each client pseudonym. The SA can therefore also see the number of messages in each queue, but not their content due to at-rest encryption.
```

Continuous monitoring adversaries have full access to the server including its volatile memory and its persisted state.

The main protection against the continuous monitoring adversary is that users interact with the services using per-group pseudonyms instead of their real identities. This means that the adversary cannot link any individual action of a user with the user's identity as seen by other users. The adversary can still see the metadata of messages, but cannot link them to the real identities of the users.

## Traffic analysis

Traffic analysis, i.e. the analysis of data traffic patterns, is not considered part of the threat model. While the use of network metadata can allow an active observer (be it on the server or just the network) to infer social graphs and other sensitive information, the Phoenix homeserver protocol does not aim to protect against such attacks. Instead, common countermeasures such as onion routing or the use of mixnets can be used in conjunction with the Phoenix homeserver protocol to mitigate such attacks. However, to the extent possible, the protocol avoids obvious traffic patterns such as batched queries, messages at regular intervals, etc.

> 

Reviewer: "active observer" is still used here but shouldn't
